When are additional setOffice calls needed?

In text-ts development I could not write text-ts unless I added an additional 
setOffice(c, "SPK") call immediately before the write attempt.
Before adding the setOffice call the exception was
Caused by: java.sql.SQLException: ORA-20048: NO_WRITE_PRIVILEGE: User doesnt have write privileges

LocationLevels just hit a similar problem and adding setOffice resolved it.  

#534 

Why are these setOffice calls now needed when they weren't needed before?
Doesn't OfficeSettingConnectionProvider  already set the session office Id for every connection?

Is it that OfficeSettingConnectionProvider is too early ?  If so, it seems like AuthDao should be setting it.

--- a/cwms-data-api/src/main/java/cwms/cda/data/dao/Dao.java
+++ b/cwms-data-api/src/main/java/cwms/cda/data/dao/Dao.java
@@ -45,13 +45,6 @@ public abstract class Dao<T> {
         return cwmsDbVersion;
     }
 
-    /**
-     * This should be called before attempting to write an object to the database.
-     * @param object Object Owned by an office id.
-     */
-    protected void setOffice(CwmsDTO object) {
-        this.setOffice(object.getOfficeId());
-    }
 
     /**
      * set session office on specific connection
@@ -68,14 +61,6 @@ public abstract class Dao<T> {
         db.setSessionOfficeId(c,office);
     }
 
-    /**
-     * Same as setOffice with DSL, however Office is known and no DTO provided.
-     * E.g. DELETE
-     * @param office
-     */
-    protected void setOffice(String office) {
-        CWMS_ENV_PACKAGE.call_SET_SESSION_OFFICE_ID(dsl.configuration(), office);
-    }
 
     public abstract List<T> getAll(Optional<String> limitToOffice);
 

--- a/cwms-data-api/src/main/java/cwms/cda/data/dao/LocationCategoryDao.java
+++ b/cwms-data-api/src/main/java/cwms/cda/data/dao/LocationCategoryDao.java
@@ -77,18 +77,24 @@ public final class LocationCategoryDao extends JooqDao<LocationCategory> {
 
     public void delete(String categoryId, boolean cascade, String office) {
         String cascadeParam = OracleTypeMap.formatBool(cascade);
-        setOffice(office);
-        CWMS_LOC_PACKAGE.call_DELETE_LOC_CAT(dsl.configuration(), categoryId, cascadeParam, office);
+        connection(dsl, conn -> {
+                    DSLContext dslContext = getDslContext(conn, office);
+                    CWMS_LOC_PACKAGE.call_DELETE_LOC_CAT(dslContext.configuration(), categoryId, cascadeParam, office);
+        });
     }
 
     public void create(LocationCategory category) {
-        setOffice(category);
-        CWMS_LOC_PACKAGE.call_CREATE_LOC_CATEGORY(dsl.configuration(), category.getId(), category.getDescription(),
-            category.getOfficeId());
+        String office = category.getOfficeId();
+        connection(dsl, conn -> {
+            DSLContext dslContext = getDslContext(conn, office);
+            CWMS_LOC_PACKAGE.call_CREATE_LOC_CATEGORY(dslContext.configuration(), category.getId(), category.getDescription(), office);
+        });
     }
 
     public void update(String oldCategoryId, String newCategoryId, String office) {
-        setOffice(office);
-        CWMS_LOC_PACKAGE.call_RENAME_LOC_CATEGORY(dsl.configuration(), oldCategoryId, newCategoryId, null, "T", office);
+        connection(dsl, conn -> {
+            DSLContext dslContext = getDslContext(conn, office);
+            CWMS_LOC_PACKAGE.call_RENAME_LOC_CATEGORY(dslContext.configuration(), oldCategoryId, newCategoryId, null, "T", office);
+        });
     }
 }

--- a/cwms-data-api/src/main/java/cwms/cda/data/dao/LocationGroupDao.java
+++ b/cwms-data-api/src/main/java/cwms/cda/data/dao/LocationGroupDao.java
@@ -43,7 +43,6 @@ import org.geojson.Feature;
 import org.geojson.FeatureCollection;
 import org.jetbrains.annotations.NotNull;
 import org.jooq.Condition;
-import org.jooq.Configuration;
 import org.jooq.DSLContext;
 import org.jooq.Field;
 import org.jooq.Record;
@@ -384,9 +383,11 @@ public final class LocationGroupDao extends JooqDao<LocationGroup> {
      * @param office The office id to use for the query.
      */
     public void delete(String categoryId, String groupId, boolean cascadeDelete, String office) {
-        setOffice(office);
-        CWMS_LOC_PACKAGE.call_DELETE_LOC_GROUP__2(dsl.configuration(), categoryId, groupId,
-            OracleTypeMap.formatBool(cascadeDelete), office);
+        connection(dsl, conn -> {
+            DSLContext dslContext = getDslContext(conn, office);
+            CWMS_LOC_PACKAGE.call_DELETE_LOC_GROUP__2(dslContext.configuration(), categoryId,
+                    groupId, OracleTypeMap.formatBool(cascadeDelete), office);
+        });
     }
 
     /**
@@ -394,13 +395,16 @@ public final class LocationGroupDao extends JooqDao<LocationGroup> {
      * @param group The location group to create.
      */
     public void create(LocationGroup group) {
-        Configuration configuration = dsl.configuration();
+        String office = group.getOfficeId();
         String categoryId = group.getLocationCategory().getId();
-        setOffice(group);
-        CWMS_LOC_PACKAGE.call_CREATE_LOC_GROUP2(configuration, categoryId,
-            group.getId(), group.getDescription(), group.getOfficeId(), group.getSharedLocAliasId(),
-            group.getSharedRefLocationId());
-        assignLocs(group);
+
+        connection(dsl, conn -> {
+            DSLContext dslContext = getDslContext(conn, office);
+            CWMS_LOC_PACKAGE.call_CREATE_LOC_GROUP2(dslContext.configuration(), categoryId,
+                    group.getId(), group.getDescription(), office, group.getSharedLocAliasId(),
+                    group.getSharedRefLocationId());
+            assignLocs(group);
+        });
     }
 
     @NotNull
@@ -416,28 +420,38 @@ public final class LocationGroupDao extends JooqDao<LocationGroup> {
      * @param newGroup The new location group id.
      */
     public void renameLocationGroup(String oldGroupId, LocationGroup newGroup) {
-        setOffice(newGroup);
-        CWMS_LOC_PACKAGE.call_RENAME_LOC_GROUP(dsl.configuration(), newGroup.getLocationCategory().getId(),
-            oldGroupId, newGroup.getId(), newGroup.getDescription(), "T", newGroup.getOfficeId());
+        String office = newGroup.getOfficeId();
+
+        connection(dsl, conn -> {
+            DSLContext dslContext = getDslContext(conn, office);
+            CWMS_LOC_PACKAGE.call_RENAME_LOC_GROUP(dslContext.configuration(), newGroup.getLocationCategory().getId(),
+                    oldGroupId, newGroup.getId(), newGroup.getDescription(), "T", office);
+        });
     }
 
     public void unassignAllLocs(LocationGroup group) {
-        setOffice(group);
-        CWMS_LOC_PACKAGE.call_UNASSIGN_LOC_GROUP(dsl.configuration(), group.getLocationCategory().getId(),
-            group.getId(), null, "T", group.getOfficeId());
+        String office = group.getOfficeId();
+        connection(dsl, conn -> {
+            DSLContext dslContext = getDslContext(conn, office);
+            CWMS_LOC_PACKAGE.call_UNASSIGN_LOC_GROUP(dslContext.configuration(), group.getLocationCategory().getId(),
+                    group.getId(), null, "T", office);
+        });
     }
 
     public void assignLocs(LocationGroup group) {
-
         List<AssignedLocation> assignedLocations = group.getAssignedLocations();
         if (assignedLocations != null) {
             List<LOC_ALIAS_TYPE3> collect = assignedLocations.stream()
                 .map(LocationGroupDao::convertToLocAliasType)
                 .collect(toList());
             LOC_ALIAS_ARRAY3 assignedLocs = new LOC_ALIAS_ARRAY3(collect);
-            setOffice(group);
-            CWMS_LOC_PACKAGE.call_ASSIGN_LOC_GROUPS3(dsl.configuration(), group.getLocationCategory().getId(),
-                group.getId(), assignedLocs, group.getOfficeId());
+
+            String office = group.getOfficeId();
+            connection(dsl, conn -> {
+                DSLContext dslContext = getDslContext(conn, office);
+                CWMS_LOC_PACKAGE.call_ASSIGN_LOC_GROUPS3(dslContext.configuration(), group.getLocationCategory().getId(),
+                        group.getId(), assignedLocs, office);
+            });
         }
     }
 }

--- a/cwms-data-api/src/main/java/cwms/cda/data/dao/TimeSeriesCategoryDao.java
+++ b/cwms-data-api/src/main/java/cwms/cda/data/dao/TimeSeriesCategoryDao.java
@@ -77,20 +77,23 @@ public class TimeSeriesCategoryDao extends JooqDao<TimeSeriesCategory> {
     }
 
     public void delete(String categoryId, boolean cascadeDelete, String office) {
-        this.setOffice(office);
-        dsl.connection((c)->
+        connection(dsl, conn -> {
+            DSLContext dslContext = getDslContext(conn, office);
             CWMS_TS_PACKAGE.call_DELETE_TS_CATEGORY(
-                getDslContext(c,office).configuration(), categoryId,
-                OracleTypeMap.formatBool(cascadeDelete), office)
-        );
+                    dslContext.configuration(), categoryId,
+                    OracleTypeMap.formatBool(cascadeDelete), office);
+        });
+
     }
 
     public void create(TimeSeriesCategory category, boolean failIfExists) {
-        this.setOffice(category);
-        dsl.connection((c) ->
+        String office = category.getOfficeId();
+
+        connection(dsl, conn -> {
+            DSLContext dslContext = getDslContext(conn, office);
             CWMS_TS_PACKAGE.call_STORE_TS_CATEGORY(
-                getDslContext(c,category.getOfficeId()).configuration(), category.getId(), category.getDescription(),
-                OracleTypeMap.formatBool(failIfExists), "T", category.getOfficeId())
-        );
+                dslContext.configuration(), category.getId(), category.getDescription(),
+                OracleTypeMap.formatBool(failIfExists), "T", office);
+        });
     }
 }

